src/background/modules/ContextMenu.js

```diff
@@ -0,0 +1,169 @@
+import * as UnicodeTransformationHandler from "/common/modules/UnicodeTransformationHandler.js"
+import * as AddonSettings from "/common/modules/AddonSettings/AddonSettings.js";
+import * as BrowserCommunication from "/common/modules/BrowserCommunication/BrowserCommunication.js";
+import { isMobile } from "/common/modules/MobileHelper.js";
+
+import { COMMUNICATION_MESSAGE_TYPE } from "/common/modules/data/BrowserCommunicationTypes.js";
+import { menuStructure, SEPARATOR_ID, TRANSFORMATION_TYPE } from "/common/modules/data/Fonts.js";
+
+const menus = browser.menus || browser.contextMenus; // fallback for Thunderbird
+const PREVIEW_STRING_CUT_LENGTH = 1000; // a setting that may improve performance by not calculating invisible parts of the context menu
+
+let lastCachedUnicodeFontSettings = null;
+
+/**
+ * Handle context menu click.
+ *
+ * @param {Object} info
+ * @param {Object} tab
+ * @returns {void}
+ * @throws {Error}
+ */
+function handleMenuChoosen(info, tab) {
+    let text = info.selectionText;
+
+    if (!text) {
+        return;
+    }
+
+    text = text.normalize();
+    const menuItem = info.menuItemId;
+    const output = UnicodeTransformationHandler.transformText(text, menuItem);
+
+    browser.tabs.executeScript(tab.id, {
+        code: `insertIntoPage("${output}");`,
+        frameId: info.frameId
+    });
+}
+
+/**
+ * Handle context menu click.
+ *
+ * @param {Object} info
+ * @param {Object} tab
+ * @returns {void}
+ * @throws {Error}
+ */
+async function handleMenuShown(info) {
+    let text = info.selectionText || "";
+
+    // do not show menu entry when no text is selected
+    if (!text) {
+        await menus.removeAll();
+        return menus.refresh();
+    }
+    // shorten preview text as it may not be shown anyway
+    if (text.length > PREVIEW_STRING_CUT_LENGTH) {
+        // to be sure, we append … anyway, in case some strange OS has a tooltip for context menus or so
+        text = `${text.substr(0, PREVIEW_STRING_CUT_LENGTH)}…`;
+    }
+    text = text.normalize();
```

Comment:
Well… my idea was to avoid work (with a potentially negative performance impact) at all on the big string, so I do this after the length check.'
Also the length is absolutely arbitrary, so it really does not matter if it's +10 or -10 chars or so. I just set a very high limit to be sure…

Reply:
In the worst case, the Unicode normalization could cut the string length in half, although I think it is fine to leave it as is. As you said, it is likely to be cut off anyway...

Reply:
Yeah, I don't know how fast that would work for… say… a 1kb string you have selected. As this function is called ~10 times or so for each menu item. If the performance is really bad, users could see a half-finished menu or an old one in such a case.

Maybe it's also premature optimization from my side though. So IMHO we should keep that issue in mind, and if someone complains we can adjust that. Otherwise let's leave at as it is…